File: server/vector/chromaClient.js
import { ChromaClient} from 'chromadb';

const chromaHost = process.env.CHROMA_HOST || "localhost";
const chromaPort = process.env.CHROMA_PORT || 8000;

const chromaClient = new ChromaClient({
  path: `http://${chromaHost}:${chromaPort}`,
});

// Create a no-op embedding function that tells Chroma we'll provide embeddings manually
class NoOpEmbeddingFunction {
  constructor() {
    this.name = "NoOpEmbeddingFunction";
  }

  async generate(texts) {
    // Return empty arrays - we'll provide embeddings manually
    return texts.map(() => []);
  }
}

export async function getProductCollection() {
  try {
    const collection = await chromaClient.getOrCreateCollection({
      name: "relief_products",
      metadata: { "hnsw:space": "cosine" },
      embeddingFunction: new NoOpEmbeddingFunction(),    // ✅ Explicitly set distance metric
    });
    
    console.log("Collection metadata:", collection.metadata);
    return collection;
  } catch (error) {
    console.error("Failed to get or create Chroma collection:", error);
    throw error;
  }
}

export async function getOrderCollection() {
  try {
    const collection = await chromaClient.getOrCreateCollection({
      name: "relief_orders",
      metadata: { "hnsw:space": "cosine" },
      embeddingFunction: new NoOpEmbeddingFunction(),    // ✅ Explicitly set distance metric
    });
    
    console.log("Collection metadata:", collection.metadata);
    return collection;
  } catch (error) {
    console.error("Failed to get or create Chroma collection:", error);
    throw error;
  }
}

export default chromaClient;

File: ai-service/chroma_client.py
import chromadb
import os
from chromadb.utils import embedding_functions

def get_chroma_collection():
    """Connect to Chroma with OpenAI embeddings enabled."""
    openai_api_key = os.getenv("OPENAI_API_KEY")
    client = chromadb.HttpClient(
        host=os.getenv("CHROMA_HOST", "localhost"),
        port=int(os.getenv("CHROMA_PORT", 8000)),
    )

    embedding_func = embedding_functions.OpenAIEmbeddingFunction(
        api_key=openai_api_key,
        model_name="text-embedding-3-small"
    )

    # Delete existing collection if it exists (one-time operation)
    # try:
    #     client.delete_collection(name="relief_products")
    #     print("Deleted existing collection")
    # except Exception as e:
    #     print(f"No existing collection to delete: {e}")

    # Create collection with explicit metadata
    collection = client.get_or_create_collection(
        name="relief_products",
        embedding_function=embedding_func,
        metadata={"hnsw:space": "cosine"}
    )
    
    print(f"Collection metadata: {collection.metadata}")
    return collection
